Metamorph for PDC

Jurisak opened the pull request:
Initial implementation

- Only query pdc metadata by component nvr knowledge
- added changes to logging_conf.py
- added initial tests

Commit:
ci_message data extractor first implementation

This plugin returns json file with important data from ci_message.
This commit contains simple test data.

Commit:
ci_message data extractor first implementation

This plugin returns json file with important data from ci_message.
This commit contains simple test data.

Commit:
ci_message data extractor first implementation

This plugin returns json file with important data from ci_message.
This commit contains simple test data.

Added travis.yml for Travis CI testing

Commit:
ci_message data extractor first implementation

This plugin returns json file with important data from ci_message.
This commit contains simple test data.

Added travis.yml for Travis CI testing

Commit:
Added travis.yml for Travis CI testing

Commit:
Added travis.yml for Travis CI testing

Commit:
Added travis.yml for Travis CI testing

Commit:
Added travis.yml for Travis CI testing

Commit:
Added travis.yml for Travis CI testing

Commit:
logging_conf renamed to support_functions

logging_conf contain multiple functions which are not related to
logging. Therefore this file was renamed.

Commit:
logging_conf renamed to support_functions

logging_conf contain multiple functions which are not related to
logging. Therefore this file was renamed.

Commit:
Documentation for  message data extractor

Commit:
Documentation for  message data extractor

Commit:
Added docstrings

Commit:
Added docstrings

Commit:
Added ansible plugin for message_data_extractor

- Still missing support for other CI-messages

Commit:
Added ansible plugin for message_data_extractor

- Still missing support for other CI-messages

Commit:
Code refactorization after code review

Commit:
Code refactorization after code review

Commit:
Initial implementation: Metamorph for PDC

- Only query pdc metadata by component nvr knowledge
- added changes to logging_conf.py
- added initial tests

Commit:
Initial implementation: Metamorph for PDC

- Only query pdc metadata by component nvr knowledge
- added changes to logging_conf.py
- added initial tests

Commit:
Initial implementation: Metamorph for PDC

- Only query pdc metadata by component nvr knowledge
- added changes to logging_conf.py
- added initial tests

Commit:
Initial implementation: Metamorph for PDC

- Only query pdc metadata by component nvr knowledge
- added changes to logging_conf.py
- added initial tests

Commit:
Initial implementation: Metamorph for PDC

- Only query pdc metadata by component nvr knowledge
- added changes to logging_conf.py
- added initial tests

Commit:
Initial implementation: Metamorph for PDC

- Only query pdc metadata by component nvr knowledge
- added changes to logging_conf.py
- added initial tests

Commit:
Added installation of requests package to .travis.yml

Commit:
Added installation of requests package to .travis.yml

Commit:
Added installation of requests package to .travis.yml

Commit:
Added installation of requests package to .travis.yml

gsr-shanks commented:
Please let 'pdc' be at the root of its JSON data.

robnester-rh commented on metamorph/plugins/morph_pdc.py:
```diff
@@ -0,0 +1,155 @@
+            response = requests.get(url, params=url_options, verify=self.ca_cert)
+            response.raise_for_status()
+            return response.json()
+        except requests.HTTPError as detail:
```
Just a quick note, you can skip escaping the double quotes in python by just starting and terminating with a single quote instead. The following line could be rewritten as:
```
logging.error('Unsuccessful connection to pdc api url '
              'with url="{0}" and options="{1}"'.format(url, url_options))

Jurisak replied:
Thanks, I didn't know about this.

robnester-rh commented on metamorph/plugins/morph_pdc.py:
```diff
@@ -0,0 +1,155 @@
+            raise PDCApiException("Unsuccessful connection to pdc api url with detail= \"{}\"".format(detail))
+
+    def setup_pdc_metadata_params(self, name, version, release):
+        for pdc_metadata_type in self.pdc_name_mapping:
```
you could simplify this as:
```
for param, param_value in self.pdc_name_mapping[pdc_metadata_type].items():
    self.pdc_name_mapping[pdc_metadata_type][param] = param_value.format(self.get_param_value(param, name, version, release))
```

Jurisak replied:
Thanks that looks good.

robnester-rh reviewed (commented):
Overall looks good. Generally, I try to provide a good docstring for each method, similar to how you've document the query_pdc method.

Jurisak commented:
Sure. I'll update the code today.

Commit:
Extracting rpm-mappings from pdc

+ added tests

Commit:
Extracting rpm-mappings from pdc

+ added tests

Commit:
Extracting rpm-mappings from pdc

+ added tests

Commit:
Extracting rpm-mappings from pdc

+ added tests

Commit:
Extracting rpm-mappings from pdc

+ added tests

Commit:
Extracting rpm-mappings from pdc

+ added tests

Commit:
Extracting rpm-mappings from pdc

+ added tests

Commit:
Extracting rpm-mappings from pdc

+ added tests

Commit:
Updated documentation

Commit:
Updated documentation

Commit:
Updated documentation

Commit:
Updated documentation

Commit:
Updated documentation

Commit:
Updated documentation

Commit:
Updated documentation

Commit:
Updated documentation

gsr-shanks commented:
Running morph_pdc.py overwrites existing metamorph.json. It should always append.

Also when we append we should have it in the following format:

```
{
  "metamorph": {
    "pdc": {
      "results": {
        "rpms": {
          "name": "bash",
          "srpm_name": "bash"
        }
      }
    },
    "resultsdb": {
      "results": {
        "tier": {
          "ci_tier": 1,
          "tier_flag": false,
          "job_name": null
        }
      }
    }
  }
}
```

Where 'metamorph' is the root element and 'pdc and resultsdb' are its branches.

Commit:
Code refactorization due code review

Commit:
Code refactorization due code review

Commit:
Code refactorization due code review

Commit:
Created pdc ansible module

+ Updated documentation

Commit:
Created pdc ansible module

+ Updated documentation

Commit:
Created pdc ansible module

+ Updated documentation

Commit:
Created pdc ansible module

+ Updated documentation

Commit:
Created pdc ansible module

+ Updated documentation

Commit:
Created pdc ansible module

+ Updated documentation

Commit:
Created pdc ansible module

+ Updated documentation

Commit:
Created pdc ansible module

+ Updated documentation

gsr-shanks commented:
@Lorquas could you please review the Ansible module for PDC ?

Commit:
Results are now appended into existing metadata.json

+ Added changes into other plugins to support this behavior

Commit:
Results are now appended into existing metadata.json

+ Added changes into other plugins to support this behavior

Commit:
Merge branch 'pdc-support' of github.com:Jurisak/metamorph into pdc-support

Jurisak commented:
@gsr-shanks Improved storing_pretty_json to append data if exists

Commit:
import fix in resultsdb files. from logging_conf -> support_functions

Commit:
import fix in resultsdb files. from logging_conf -> support_functions

Lorquas commented on .travis.yml:
```diff
@@ -6,5 +6,6 @@ python:
   - "3.6"
 install:
   - python setup.py install
+  - pip install requests
```
I envision that this and other PRs will pip install many things (I've started to see it in other PRs).  You can put these in a `requirements.txt` file and do `pip install -r reqirements.txt`  to keep things from getting wordy.  If you don't want to maintain the requirements in multiple places, and only in the `setup.py` you can have the requirements file like this: 

```
~/P/metamorph ❯❯❯ cat requirements.txt
.
~/P/metamorph ❯❯❯
```

and then `pip install -r requirements.txt` will install all the deps in setup.py's `install_requires`. 
Good article on this: https://caremad.io/posts/2013/07/setup-vs-requirement/

Jurisak replied:
Ok that's a good idea. Thanks.

Lorquas commented on metamorph/lib/support_functions.py:
```diff
@@ -0,0 +1,45 @@
+
+
+class ExistingMetamorphMetadataException(Exception):
+    pass
```
So here we have an exception with the word Exception in the name (redundant), but it also doesn't do anything.  We all know that the thing that comes right after `raise` must be an exception, and that any human reading a traceback is going to read the exception message anyway (which should always be verbose and full of good information).  We also are not using this exception in any special way programatically or catching this ever in any special use case.  We should be using `LookupError`  because its built in, doesn't complicate the code any further, and everybody using python knows what a LookupError is.

Jurisak replied:
My original thought was to create custom exceptions which can be handled afterwards, but they are not now ...

Lorquas replied:
Yeah in that case they're fine, hard to make these reviews based on future use

Lorquas commented on metamorph/lib/support_functions.py:
```diff
@@ -0,0 +1,45 @@
+            with open(output, 'w') as metamorph:
+                json.dump(existing_metadata, metamorph, indent=2)
+        else:
+            raise ExistingMetamorphMetadataException("ERROR: Wrong format of given '{}'. "
```
Should be LookupError, see note above.

Lorquas commented on metamorph/library/pdc.py:
```diff
@@ -0,0 +1,303 @@
+
+
+class PDCApiException(Exception):
+    pass
```
This usage of Exception here is a little more interesting, as we're using it here to re-raise inbound exceptions with a different message.  We are not, however, ever catching this custom exception. This can be a little cleaner in python3 with something like:

```
except Exception as e:
        import sys
        raise type(e)(str(e) +
                      ' some custom message here: %s' % someArg).with_traceback(sys.exc_info()[2])
```
which has the benefit over our solution as the traceback tree will be more direct to the actual exception.  It can also be used to wrap the occasional `requests.ConnectionError` (in addition to `HTTPError`) that will get thrown while maintaining the same context.

The problem with this approach is that it doesn't work with python2 syntax.  We would have to make use of `six.reraise`, and throw some [boilerplate](http://stackoverflow.com/a/38772267) around that to handle both python2 and 3.  At this point we've added some level of complication all in the quest to reduce the traceback by a few lines --  I'm not sure if my suggestion is worth it.

I know we have versions 2.7 - 3.6 in the travis.yml, but has there been an actual decision on which python versions we're supporting?  Dropping 2.7 will give us a lot of flexibility with newer features, but might limit certain deployment situations (which are yet undefined?)

Jurisak replied:
@Lorquas How about to discuss python2.7 support on the next Metamorph meeting?

Lorquas replied:
Python 3 it !

Lorquas reviewed (commented):
Overall this is very good.  In lieu of substantive changes to this PR directly, I've opted to pontificate on exception handling philosophy.  We may choose to ignore depending on how fast and loose we want to play.

I do have some greater questions on the structure of this project but that lies outside the scope of this PR.

LGTM; shipit

Lorquas commented on setup.py:
```diff
@@ -8,5 +8,5 @@
     author='Jiri Kulda',
     author_email='Kulda12@seznam.cz, jkulda@redhat.com',
-    description='Exhibiting structural change of test metadata.', requires=['ansible']
+    description='Exhibiting structural change of test metadata.', requires=['ansible', 'requests']
```
requires should be on its own line like all the other kwargs, no?

also isn't it `install_requires` ? (it is if you want to use that requirements.txt trick above)

Lorquas committed:
Update README.md

added travis build badge.

Commit:
Added requirements.txt file support

Commit:
Fixed problem with printing json results in messagehub plugin

Plugin does not printed json file in pretty print format.
Fixed by adding parsing by json package.

+ added tests

Commit:
Fixed problem with printing json results in messagehub plugin

Plugin does not printed json file in pretty print format.
Fixed by adding parsing by json package.

+ added tests

Commit:
Fixed problem with printing json results in messagehub plugin

Plugin does not printed json file in pretty print format.
Fixed by adding parsing by json package.

+ added tests

Commit:
Fixed problem with printing json results in messagehub plugin

Plugin does not printed json file in pretty print format.
Fixed by adding parsing by json package.

+ added tests

Commit:
Added package update to .travis.yml

Commit:
Added package update to .travis.yml

Commit:
Added package update to .travis.yml

Commit:
Added package update to .travis.yml

Commit:
Added package update to .travis.yml

Commit:
Added test for pdc plugin and ansible version

Commit:
Added test for pdc plugin and ansible version

Commit:
Added test for pdc plugin and ansible version

Commit:
Added test for pdc plugin and ansible version

Commit:
Added tests for dispatcher plugin.

+ ansible plugin tests

Commit:
Added tests for dispatcher plugin.

+ ansible plugin tests

Commit:
Renamed job_names option to job_name

job_names option was changed to job_name because job_names will
always contain only singular values

! Should not be merged before resultsdb-updater PR #26

robnester-rh reviewed (commented):
+1 to @Lorquas comments regarding custom exceptions. Unless there is a compelling case for them, I'd recommend using the exceptions in the standard library.

gsr-shanks commented:
Please resolve conflicts and comments regarding custom exceptions.

Jurisak commented:
@Lorquas Could you please create Issue describing desired exception logic?

Lorquas commented:
@Jurisak will do, We could probably merge this then and handle the exceptions later in that issue...

Commit:
Merge branch 'pdc-support' of github.com:Jurisak/metamorph into pdc-support

Jurisak commented:
@Lorquas yes. I agree.

Jurisak commented:
FYI: Somebody cleaned data from release-components field. For some components it is not working. Therefore rpm-mapping is not working as well.

gsr-shanks commented:
Please resolve conflicts.

Commit:
Merge branch 'master' of github.com:Jurisak/metamorph into pdc-support

Jurisak commented:
@gsr-shanks waiting for merge

Commit:
PDC plugins are now inheriting from MetamorphPlugin class
